llvm/lib/Transforms/Yk/ControlPoint.cpp

```diff
@@ -88,36 +89,149 @@ CallInst *findControlPointCall(Module &M) {
+      // instructions of all successor blocks.
+      for (unsigned SuccIdx = 0; SuccIdx < Term->getNumSuccessors(); SuccIdx++)
+        SuccInsts.insert(&*Term->getSuccessor(SuccIdx)->begin());
+    }
+    return SuccInsts;
   }
 
-  // Then add anything which dominates the control point to the live set.
-  for (auto &BB : *Func) {
-    if (!DT.dominates(cast<Instruction>(OldCtrlPoint), &BB)) {
-      for (auto &I : BB) {
-        if ((!I.getType()->isVoidTy()) &&
-            (DT.dominates(&I, cast<Instruction>(OldCtrlPoint)))) {
-          Vec.push_back(&I);
+  /// Replaces the value set behind the pointer `S` with the value set `R` and
+  /// returns whether the set behind `S` changed.
+  bool updateValueSet(std::set<Value *> *S, const std::set<Value *> R) {
+    bool Changed = (*S != R);
+    *S = R;
+    return Changed;
+  }
+
+public:
+  LivenessAnalysis(Function *Func) {
+    // Compute defs and uses for each instruction.
+    std::map<Instruction *, std::set<Value *>> Defs;
+    std::map<Instruction *, std::set<Value *>> Uses;
+    for (BasicBlock &BB : *Func) {
+      for (Instruction &I : BB) {
+        // Record what this instruction defines.
+        if (!I.getType()->isVoidTy())
+          Defs[&I].insert(cast<Value>(&I));
+
+        // Record what this instruction uses.
+        //
+        // Note that Phi nodes are special and must be skipped. If we consider
+        // their operands as uses, then Phi nodes in loops may use variables
+        // before they are defined, and this messes with the algorithm.
+        //
+        // The book doesn't cover this quirk, as it explains liveness for
+        // non-SSA form, and thus doesn't need to worry about Phi nodes.
+        if (isa<PHINode>(I))
+          continue;
+
+        for (auto *U = I.op_begin(); U < I.op_end(); U++) {
+          if ((!isa<Constant>(U)) && (!isa<BasicBlock>(U)) &&
+              (!isa<MetadataAsValue>(U)) && (!isa<InlineAsm>(U))) {
+            Uses[&I].insert(*U);
+          }
         }
       }
     }
+
+    // A function implicitly defines its arguments.
+    //
+    // To propagate the arguments properly we pretend that the first instruction
+    // in the entry block defines the arguments.
+    Instruction *FirstInst = &*Func->getEntryBlock().begin();
+    for (auto &Arg : Func->args())
+      Defs[FirstInst].insert(&Arg);
+
+    // Compute the live sets for each instruction.
+    //
+    // This is the fix-point of the following data-flow equations (page 206 in
+    // the book referenced above):
+    //
+    //    in[I] = use[I] ∪ (out[I] - def[I])
+    //
+    //    out[I] =       ∪
+    //             (S in succ[I])    in[S]
+    //
+    // Note that only the `Ins` map is kept after this constructor ends, so
+    // only `Ins` is a field.
+    std::map<Instruction *, std::set<Value *>> Outs;
+    bool Changed;
+    do {
+      Changed = false;
+      // As the book explains, fix-points are reached quicker if we process
```

Comment:
Could you give a little hint as to why reversing direction helps? [I can guess, but it never hurts to quickly explain it!]

Reply:
My intuition is that it's because the algorithm introduces liveness as use-sites and propagates them backwards until they are killed at their def-sites.

If you compute the fixed-point forwards then in each iteration of the algorithm you can only propagate liveness backwards by at most one instruction.

If you do it backwards, you can have a whole run of backwards propagation in just on iteration of the algorithm.

Reply:
Here's what the book says:

![image](https://user-images.githubusercontent.com/604955/162191872-7c18d973-d367-45c3-9c4b-53bb04bea8a3.png)

Also I've just realised that if we swap the order of computing ins and outs, then we might fix sooner too. The book mentions this only in passing!

![image](https://user-images.githubusercontent.com/604955/162192096-410d9a1b-3d1c-4a6f-83d0-c72afe2e4247.png)

Investigating...

Reply:
Indeed. For the `simple_interp_loop2` test, we can fix in 3 iterations instead of in 67!

Fixed in 11a44355a0090a0068d9ff0090d0287dd4ab901a

Reply:
Blimey -- that wasn't the answer I was expecting but I like it!